sessionctx/variable/session.go

```diff
@@ -549,6 +549,36 @@ func validateReadConsistencyLevel(val string) error {
 	}
 }
 
+// SetUserVarVal set user defined variables' value
+func (s *SessionVars) SetUserVarVal(name string, dt types.Datum) {
+	s.userVarLock.Lock()
+	defer s.userVarLock.Unlock()
+	s.userVarValues[name] = dt
+}
+
+// GetUserVarVal get user defined variables' value
+func (s *SessionVars) GetUserVarVal(name string) (types.Datum, bool) {
+	s.userVarLock.RLock()
+	defer s.userVarLock.RUnlock()
+	dt, ok := s.userVarValues[name]
+	return dt, ok
+}
+
+// SetUserVarType set user defined variables' type
+func (s *SessionVars) SetUserVarType(name string, fd *types.FieldType) {
```

Comment:
Maybe it's better to rename `fd` to `ft`. Now we use the `fd` to indicate the function dependency. The following are same.